### PR TITLE
Configure the check-group app

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

A check run status is either required or not on GitHub.  It can't be marked conditionally. It can run conditionally, but even if it's not run on a PR, it would still be required if it's required on any part of the repos code.

A solution is to have a single required job that checks what jobs should be required based on conditions. This marketplace app does exactly that.

This should improve time-to-merge and reduce CI overhead

After this is merged only the following jobs would be marked as required:
- Lightning CI required checker
- WIP
- pre-commit.ci
- Milestone Check
- Schema check

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [n/a] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .github/checkgroup.yml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      - "examples/run_ddp_examples.sh"
+      - "examples/run_pl_examples.sh"
+      - "requirements/pytorch/**"
+      - "src/pytorch_lightning/**"
+      - "tests/tests_pytorch/**"
+      - "setup.cfg"  # includes pytest config
+      - "pyproject.toml"  # includes mypy config
+      - ".readthedocs.yml"
+      - ".github/workflows/_check-schema.yml"
+      - ".github/workflows/ci-schema.yml"
```
Comment (Borda) [inline .github/checkgroup.yml:19]: I would keep this as always required, regardless job, they are also 1min or so

File: .github/checkgroup.yml [hunk 2]
```diff
@@ -0,0 +1,134 @@
+custom_service_name: "Lightning CI required checker"
+subprojects:
+  - id: "pytorch_lightning"
```
Comment (Borda) [inline .github/checkgroup.yml:3]: Shall we also separate docs for each as changes of docs do not need to run full code testing...

**Review by tchaton — APPROVED** [review]
LGTM !

File: .github/checkgroup.yml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    paths:
+      - ".circleci/**"
```
Comment (Borda) [inline .github/checkgroup.yml:5]: Maybe also as an additional split, if we change something only in GPU test pipeline, we do not need to run all CPU testing, right?

**Review by Borda — APPROVED** [review]

File: .github/checkgroup.yml (hunk 2 above)
Comment (carmocca) [reply .github/checkgroup.yml:3]: Correct, let's do that

File: .github/checkgroup.yml (hunk 1 above)
Comment (carmocca) [reply .github/checkgroup.yml:25]: Sure

File: .github/checkgroup.yml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  - id: "install"
+    paths:
+      - ".actions/setup_tools.py"
+      - ".github/workflows/ci-pkg-install.yml"
+      - "setup.py"
+      - "src/lightning/**"
+      # all __about__, __version__, __setup__
+      - "src/*/__*.py"
+    checks:
+      - "install-meta-pypi (macOS-11, 3.8)"
```
Comment (Borda) [inline .github/checkgroup.yml:153]:
```suggestion
      - "install-meta-pypi (*)"
```
Do you think that reg expression also works here?

Comment (carmocca) [reply .github/checkgroup.yml:153]: I have no idea. But I suggest doing this experimentation in a follow up once we have the app running

File: .github/checkgroup.yml [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      - "Cloud Test (drive)"
+      - "Cloud Test (idle_timeout)"
+      - "Cloud Test (payload)"
+      - "Cloud Test (template_jupyterlab)"
+      - "Cloud Test (template_react_ui)"
+      - "Cloud Test (template_streamlit_ui)"
+      - "Cloud Test (v0_app)"
+      - "doctest (app)"
+      - "make-docs (app)"
+      - "pytest (macOS-11, 3.8, latest)"
```
Comment (carmocca) [inline .github/checkgroup.yml:128]: As a note, we'll have to be mindful in the future about job name collisions.

If i'm not mistaken, these are app pytest tests, but the names do not make it clear

File: .github/checkgroup.yml (hunk 4 above)
Comment (Borda) [reply .github/checkgroup.yml:153]: just for visibility: https://github.com/tianhaoz95/check-group/issues/1078

File: .github/checkgroup.yml [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      - "examples/pl_loops/mnist_lite.py"
+      - "examples/pl_fault_tolerant/automatic.py"
+      - "examples/test_pl_examples.py"
+      - "examples/pl_integrations/dali_image_classifier.py"
+      - "requirements/pytorch/**"
+      - "src/pytorch_lightning/**"
+      - "tests/tests_pytorch/**"
+      - "setup.cfg"  # includes pytest config
+      - "pyproject.toml"  # includes mypy config
+      - ".readthedocs.yml"
```
Comment (Borda) [inline .github/checkgroup.yml:35]:
```suggestion
```
there is no check for this

File: .github/checkgroup.yml [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      - "examples/pl_basics/autoencoder.py"
+      - "examples/pl_loops/mnist_lite.py"
+      - "examples/pl_fault_tolerant/automatic.py"
+      - "examples/test_pl_examples.py"
+      - "examples/pl_integrations/dali_image_classifier.py"
+      - "requirements/pytorch/**"
+      - "src/pytorch_lightning/**"
+      - "tests/tests_pytorch/**"
+      - "setup.cfg"  # includes pytest config
+      - "pyproject.toml"  # includes mypy config
```
Comment (Borda) [inline .github/checkgroup.yml:34]:
```suggestion
```
you are already running my as extra group

File: .github/checkgroup.yml [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+      - "build-NGC"
+      - "build-pl (3.9, 1.10, 11.3.1)"
+      - "build-pl (3.9, 1.11, 11.3.1)"
+      - "build-pl (3.9, 1.12, 11.3.1)"
+      - "build-pl (3.9, 1.9, 11.1.1)"
+      - "build-xla (3.7, 1.12)"
+
+  - id: "pytorch_lightning: mypy"
+    paths:
+      - ".github/workflows/code-checks.yml"
```
Comment (Borda) [inline .github/checkgroup.yml:98]:
```suggestion
      - ".github/workflows/code-checks.yml"
      - "pyproject.toml"  # includes mypy config
```

File: .github/checkgroup.yml [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    checks:
+      - "mypy"
+
+  - id: "lightning_app"
+    paths:
+      - ".github/workflows/ci-app*.yml"
+      - ".github/workflows/docs-*.yml"
```
Comment (Borda) [inline .github/checkgroup.yml:105]:
```suggestion
```
Moved to Docs check bellow

File: .github/checkgroup.yml (hunk 6 above)
Comment (carmocca) [reply .github/checkgroup.yml:35]: But a change to this file could break the docs build, no?

File: .github/checkgroup.yml (hunk 7 above)
Comment (carmocca) [reply .github/checkgroup.yml:34]: I guess we can move the file path to the mypy group

File: .github/checkgroup.yml (hunk 6 above)
Comment (Borda) [reply .github/checkgroup.yml:34]: no, this file is used only in RTFD, sot the break will be visible on deploy not on CI/PR

**Pull request merged**